apisix/healthcheck_manager.lua

```diff
@@ -0,0 +1,288 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local require = require
+local ipairs   = ipairs
+local pcall   = pcall
+local exiting      = ngx.worker.exiting
+local pairs    = pairs
+local tostring = tostring
+local core = require("apisix.core")
+local config_local   = require("apisix.core.config_local")
+local healthcheck
+local events = require("apisix.events")
+local tab_clone = core.table.clone
+local timer_every = ngx.timer.every
+local _M = {
+    working_pool = {},     -- resource_path -> {version = ver, checker = checker}
+    waiting_pool = {}      -- resource_path -> resource_ver
+}
+local healthcheck_shdict_name = "upstream-healthcheck"
+local is_http = ngx.config.subsystem == "http"
+if not is_http then
+    healthcheck_shdict_name = healthcheck_shdict_name .. "-" .. ngx.config.subsystem
+end
+
+
+local function get_healthchecker_name(value)
+    return "upstream#" .. (value.resource_key or value.upstream.resource_key)
+end
+_M.get_healthchecker_name = get_healthchecker_name
+
+
+local function fetch_latest_conf(resource_path)
+    local resource_type, id
+    -- Handle both formats:
+    -- 1. /apisix/<resource_type>/<id>
+    -- 2. /<resource_type>/<id>
+    if resource_path:find("^/apisix/") then
+        resource_type, id = resource_path:match("^/apisix/([^/]+)/([^/]+)$")
+    else
+        resource_type, id = resource_path:match("^/([^/]+)/([^/]+)$")
+    end
```

Comment:
why we still need to handle those two format?

Reply:
We will have to handle these two cases anyways either here or in filter_upstream to construct a resource key by stripping the "/apisix". And even then inside this function we will have to break down /routes/1 into "routes" and 1 to first data = fetch_created_obj(routes) and then data:get(id). 
So I thought it will be better to have all the key construction logic and fetching logic at one place for easier to read.
Do you think there is advantage of moving this to filter_upstream?

Reply:
Either based on matching like this, or explicitly using resource_type in filter_upstream, we will have to handle these two cases. I picked here to have all key generation logic + fetching data from generated key logic at one place.